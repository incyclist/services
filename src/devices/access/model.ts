import { IncyclistCapability} from "incyclist-devices"


export type InterfaceState = 'connected'| 'disconnected' | 'unknown' | 'connecting' | 'disconnecting' | 'unavailable'
export type ScanState = 'start-requested' | 'started' | 'stop-requested' | 'stopped' | 'idle'

/*
 * Provides the information about a single Interface
*/
export interface InterfaceInfo {
    /** name of the interface */
    name:string

    /**  is the interface enabled by the user/system */
    enabled:boolean

    /** connection state, only when state is 'connected', the interface is ready to be used  */
    state: InterfaceState

    /** provides information if the interface is currently performing a scan */
    isScanning: boolean;

    /** additional properties provided to the Interface ( e.g. timeouts, protocol for SerialInterface, ...) */
    properties?:InterfaceAccessProps
}

/*
 * Additional information that can be provided for an Interface
 *
 * This can be provided as default for all interfaces, or for every interface individually. 
 * The individual settings overwrite the default.
*/

export interface InterfaceAccessProps {
    /** Timeout for a connect attempt */
    connectTimeout?: number;

    /** Timeout for a scan attempt */
    scanTimeout?:number;

    /** TCP Port (only relevant for 'tcpip' interface) */
    port?: number

    /** Protocol to be used (only relevant for 'tcpip' and 'serial' interface) */
    protocol?: string;

    /** If set to `true` the service will continously try to connect to this interface, otherwise [[connect]] needs to be explicitely called */
    autoConnect?: boolean
}


export interface InterfaceList {[index: string]: InterfaceInfo}


export interface ScanFilter {
    interfaces?: string[],
    capabilities?: (IncyclistCapability|'bike')[]
    protocol?: string
    protocols?: string[]
    profile?: string
    excludeDisabled?: boolean
}


