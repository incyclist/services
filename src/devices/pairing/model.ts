import { DeviceData, DeviceSettings, IncyclistCapability } from "incyclist-devices"
import { EnrichedInterfaceSetting } from "../access"
import { AdapterRideInfo, RideServiceCheckFilter } from "../ride"


export interface DevicePairingData {
    udid:string,
    name:string,
    interface:string,
    connectState?:string,
    value?:number,
    unit?:string
    selected:boolean
}

export interface CapabilityData {
    capability: IncyclistCapability,
    value?: number,
    unit?: string,
    connectState?: DevicePairingStatus,   
    deviceName:string
    deviceNames: string
    selected:string
    disabled:boolean
    interface:string
    devices:Array<DevicePairingData>
}




export interface PairingData {
    udid: string,
    data: Array<CapabilityData>
}

export interface PairingProps {
    filter?: RideServiceCheckFilter,
    enforcedScan?: boolean
}

export interface PairingInfo {
    promiseCheck?: Promise<boolean>,
    promiseStart?: Promise<boolean>,
    promiseScan?:Promise<DeviceSettings[]>
    props:PairingProps
    data?:Array< {udid:string, data:DeviceData, ts:number}>
}

export interface PairingState {
    capabilities?: Array<CapabilityData>
    interfaces?: Array<EnrichedInterfaceSetting>
    canStartRide?:boolean    
    adapters?: Array<AdapterRideInfo>
}

export interface DeviceSelectState {
    capability:IncyclistCapability
    devices:Array<DevicePairingData>
}



export interface PairingSettings {
    onStateChanged?: (newState:PairingState)=>void
    onDeviceSelectStateChanged?:(newState:DeviceSelectState)=>void
    capabilityForScan?: IncyclistCapability
}



export type DevicePairingStatus = 'connecting'|'connected'|'failed'|'waiting'|'paused'