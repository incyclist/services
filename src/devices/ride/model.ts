import { DeviceProperties } from "incyclist-devices/lib/types/device";
import { AdapterInfo, ExtendedIncyclistCapability } from "../configuration";

export interface AdapterRideInfo extends AdapterInfo {
    isStarted: boolean
    tsLastData?: number
    isHealthy?: boolean
    ivToCheck?: NodeJS.Timeout

}

export interface AdapterStateInfo {
    udid: string,
    name:string,
    isControl:boolean,
    capabilities:Array<ExtendedIncyclistCapability>
    isStarted:boolean
}



export interface RideServiceDeviceProperties extends DeviceProperties {
    forceErgMode?: boolean
    startPos?: number,
    realityFactor?:number,
    rideMode?:string
    route?
}

export interface RideServiceCheckFilter {
    interface?: string
    interfaces?: string[]
    udid?: string
}

export interface Point {
    elevation: number,
    distance: number
}

export interface PreparedRoute {
    name: string
    description: string
    programId: number
    type:  string
    totalDistance:number,
    lapMode: boolean,
    minElevation: number,
    maxElevation: number,
    sampleRate:number,
    points: Point[]    
}

