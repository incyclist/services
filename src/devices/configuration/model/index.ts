export * from './app'
export * from './repository'
export * from './service'