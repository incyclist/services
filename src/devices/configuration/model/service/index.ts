import { CyclingMode, IncyclistDeviceAdapter } from "incyclist-devices";
import { ExtendedIncyclistCapability } from "../app";

export interface AdapterInfo {
    udid: string
    adapter: IncyclistDeviceAdapter
    capabilities: ExtendedIncyclistCapability[]
}

export interface DeviceModeInfo {
    udid: string
    mode: string
    isERG?: boolean,
    isSIM?: boolean
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    settings: any
    options?: CyclingMode[]
}