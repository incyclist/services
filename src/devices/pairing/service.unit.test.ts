/* eslint-disable @typescript-eslint/no-explicit-any */
import { DeviceSettings, IncyclistCapability } from 'incyclist-devices'
import { CapabilityInformation, DeviceConfigurationService,} from '../configuration'
import {DevicePairingService, mappedCapability} from './service'
import clone from '../../utils/clone'
import { DeviceAccessService, useDeviceAccess } from '../access'
import { DeviceRideService, useDeviceRide } from '../ride'
import UserSettingsMock from '../../settings/user/mock'
import { DevicePairingStatus, PairingState } from './model'
import { UserSettingsService } from '../../settings'

class TestWrapper extends DevicePairingService {
    
    _get(member:string){
        return (this as any)[member]
        
    }
    _set(member:string,value) {
        (this as any)[member] = value
        //console.log(this)
    }

    _call(func:string,...args) {
        const fn = (this as any)[func].bind(this)
        
        try {
            return fn(...args)
        }
        catch(err) {
            console.log(func,'->',fn,err.message)
        }
    }
}

describe('PairingService',()=>{

    describe('api',()=>{
        describe('start',()=>{
            class TestWrapper extends DevicePairingService {
                getState() {
                    return this.state
                }
                getPairingRetryDelay() {
                    return 10;
                }
            }

            let svc:TestWrapper
            let userSettings:UserSettingsMock
            let ride:DeviceRideService
            let access:DeviceAccessService
            
            let logEvent
            let settings:any = {
                interfaces: [
                    { name:'ant', enabled:true },
                    { name:'ble', enabled:true },
                    { name:'serial', enabled:false, protocol:'Daum Classic' },
                    { name:'tcpip', enabled:true },
                ],
                "NEW_UI":true

            }

            beforeEach( ()=>{
                userSettings = new UserSettingsMock(settings)
                ride = useDeviceRide()
                access = useDeviceAccess()
                
                svc = new TestWrapper()

                ride.startAdapters=jest.fn()
                ride.getAdapters=jest.fn().mockReturnValue([])
                ride.cancelStart = jest.fn()

                access.scan = jest.fn()
                access.enableInterface = jest.fn()
                access.stopScan = jest.fn()
                access.enrichWithAccessState = jest.fn( (interfaces)=>{ return interfaces.map(i => ({...i, state:'connected', isScanning:false}))})

                logEvent  = jest.spyOn(svc as any,'logEvent')
            })

            afterEach( ()=>{
                (DeviceAccessService as any)._instance = undefined;
                (DeviceConfigurationService as any)._instance = undefined;
                (DeviceRideService as any)._instance = undefined;  
                (UserSettingsService as any)._instance = undefined              
            })

            test('no devices in configuration will start scan',async ()=>{
                settings.devices = []
                settings.cabilities= {}

                let updates: Array<PairingState> = [];
                const res = await new Promise (done => {
                    svc.start( (status:PairingState)=>{ updates.push(status);if (updates.length==2) done(updates)  })
                })

                expect(res).toMatchSnapshot()        
                expect(logEvent).toHaveBeenCalledWith( expect.objectContaining( {message:'Start Scanning'}))
                expect(ride.startAdapters).not.toHaveBeenCalled()
                expect(access.scan).toHaveBeenCalled()
            })

            test('no devices in configuration; FE device found in scan',async ()=>{
                settings.devices = []
                settings.cabilities= {}

                const device = {interface:'ant', profile: "FE", deviceID: 1234 }
                userSettings.settings = {}
                access.scan = jest.fn( async ()=>{
                    access.emit('device', device)
                    
                    return [device]
                })
                let updates: Array<PairingState> = [];
                let numLogs = 0;
                // we are expecting 3 status updates, the last one to contain the 
                const res = await new Promise (done => {
                    svc.on('log',()=>{ 
                        ++numLogs
                        if (numLogs===3) 
                            done(updates)
                    })                   
                    svc.start( (status:PairingState)=>{ 
                        updates.push(status)
                    })
                })

                //expect(res).toMatchSnapshot()        
                //console.log(JSON.stringify(res,undefined,2))

                expect(logEvent).toHaveBeenNthCalledWith( 1,expect.objectContaining( {message:'Start Scanning'}))
                expect(logEvent).toHaveBeenNthCalledWith( 2,expect.objectContaining( {message:'device detected', device:{deviceID:1234, interface:'ant', profile:'FE'}}))
                expect(logEvent).toHaveBeenNthCalledWith( 3,expect.objectContaining( {message:'Start Pairing'}))
                expect(access.scan).toHaveBeenCalled()
                expect(ride.startAdapters).toHaveBeenCalled()
                expect(svc.getState().canStartRide).toBe(true)
                expect(svc.getState().capabilities?.find(c=>c.capability===IncyclistCapability.Control)).toMatchObject({deviceName:'Ant+FE 1234'})
                expect(svc.getState().capabilities?.find(c=>c.capability===IncyclistCapability.Power)).toMatchObject({deviceName:'Ant+FE 1234'})
                expect(svc.getState().capabilities?.find(c=>c.capability===IncyclistCapability.Speed)).toMatchObject({deviceName:'Ant+FE 1234'})
                expect(svc.getState().capabilities?.find(c=>c.capability===IncyclistCapability.Cadence)).toMatchObject({deviceName:'Ant+FE 1234'})
                expect(svc.getState().capabilities?.find(c=>c.capability===IncyclistCapability.HeartRate)).toMatchObject({deviceName:undefined})
                expect(svc.getState).toMatchSnapshot()

            })

            test('no devices in configuration; HRM sensor found in scan',async ()=>{

                const device = {interface:'ant', profile: "HR", deviceID: 1234 }
                settings.devices = []
                settings.cabilities= {}
                
                access.scan = jest.fn( async ()=>{
                    access.emit('device', device)                    
                    return [device]
                })

                let updates: Array<PairingState> = [];
                let numLogs = 0;
                // we are expecting 3 status updates, the last one to contain the 
                const res = await new Promise (done => {
                    svc.on('log',()=>{ 
                        ++numLogs
                        if (numLogs===3) 
                            done(updates)
                    })                   
                    svc.start( (status:PairingState)=>{ 
                        updates.push(status)
                    })
                })

                expect(logEvent).toHaveBeenNthCalledWith( 1,expect.objectContaining( {message:'Start Scanning'}))
                expect(logEvent).toHaveBeenNthCalledWith( 2,expect.objectContaining( {message:'device detected', device:{deviceID:1234, interface:'ant', profile:'HR'}}))
                expect(logEvent).toHaveBeenNthCalledWith( 3,expect.objectContaining( {message:'Start Scanning'}))
                expect(access.scan).toHaveBeenCalled()
                expect(ride.startAdapters).not.toHaveBeenCalled()
                expect(svc.getState().canStartRide).toBe(false)
                expect(svc.getState().capabilities?.find(c=>c.capability===IncyclistCapability.Control)).toMatchObject({deviceName:undefined})
                expect(svc.getState().capabilities?.find(c=>c.capability===IncyclistCapability.Power)).toMatchObject({deviceName:undefined})
                expect(svc.getState().capabilities?.find(c=>c.capability===IncyclistCapability.Speed)).toMatchObject({deviceName:undefined})
                expect(svc.getState().capabilities?.find(c=>c.capability===IncyclistCapability.Cadence)).toMatchObject({deviceName:undefined})
                expect(svc.getState().capabilities?.find(c=>c.capability===IncyclistCapability.HeartRate)).toMatchObject({deviceName:'Ant+HR 1234'})
                expect(svc.getState).toMatchSnapshot()

            })

            test('FE device in configuration',async ()=>{

                settings.devices = [
                        { "udid": "1", "settings": {"profile": "FE", "interface": "ant","deviceID": 1234 } }
                    ];

                settings.capabilities = [
                        {"capability": "control","devices": ["1"],"selected": "1","disabled": false },
                        {"capability": "power","devices": ["1"],"selected": "1","disabled": false },
                        {"capability": "cadence","devices": ["1"],"selected": "1","disabled": false },
                        {"capability": "speed","devices": ["1"],"selected": "1","disabled": false },
                        {"capability": "heartrate","devices": [],"selected": undefined,"disabled": false },
                    ]
                
                ride.startAdapters=jest.fn( async ()=> { 
                    ride.emit('pairing-start',)
                    return false
                })
                ride.getAdapters= jest.fn().mockReturnValue( 
                    [ { udid:'1', 
                        capabilties:[IncyclistCapability.Control, IncyclistCapability.Power, IncyclistCapability.Speed, IncyclistCapability.Cadence], 
                        adapter: {  
                            isStarted:jest.fn().mockReturnValue(false),
                            pause:jest.fn().mockResolvedValue(true)
                        }
                      }])

                

                let updates: Array<PairingState> = [];
                let numLogs = 0;
                // we are expecting 3 status updates, the last one to contain the 
                const res = await new Promise (done => {
                    svc.on('log',()=>{ 
                        ++numLogs
                        if (numLogs===2) 
                            done(updates)
                    })                   
                    svc.start( (status:PairingState)=>{ 
                        updates.push(status)
                    })
                })

                //expect(res).toMatchSnapshot()        
                //console.log(JSON.stringify(res,undefined,2))

                expect(logEvent).toHaveBeenNthCalledWith( 1,expect.objectContaining( {message:'Start Pairing'}))
                expect(access.scan).not.toHaveBeenCalled()
                expect(ride.startAdapters).toHaveBeenCalled()

                expect(svc.getState().canStartRide).toBe(true)
                expect(svc.getState().capabilities?.find(c=>c.capability===IncyclistCapability.Control)).toMatchObject({deviceName:'Ant+FE 1234', connectState:'connecting'})
                expect(svc.getState().capabilities?.find(c=>c.capability===IncyclistCapability.Power)).toMatchObject({deviceName:'Ant+FE 1234', connectState:'connecting'})
                expect(svc.getState().capabilities?.find(c=>c.capability===IncyclistCapability.Speed)).toMatchObject({deviceName:'Ant+FE 1234', connectState:'connecting'})
                expect(svc.getState().capabilities?.find(c=>c.capability===IncyclistCapability.Cadence)).toMatchObject({deviceName:'Ant+FE 1234', connectState:'connecting'})
                expect(svc.getState().capabilities?.find(c=>c.capability===IncyclistCapability.HeartRate)).toMatchObject({deviceName:undefined})
                expect(svc.getState).toMatchSnapshot()

            })

            test.skip('FE and PWR device in configuration, PWR device fails to start',async ()=>{ 
                settings.devices = [
                    { "udid": "1", "settings": {"profile": "FE", "interface": "ant","deviceID": 1234 } },
                    { "udid": "2", "settings": {"profile": "PWR", "interface": "ant","deviceID": 2345 } }
                ];

            settings.capabilities = [
                    {"capability": "control","devices": ["1"],"selected": "1","disabled": false },
                    {"capability": "power","devices": ["1","2"],"selected": "2","disabled": false },
                    {"capability": "cadence","devices": ["1"],"selected": "1","disabled": false },
                    {"capability": "speed","devices": ["1"],"selected": "1","disabled": false },
                    {"capability": "heartrate","devices": [],"selected": undefined,"disabled": false },
                ]
            
            ride.startAdapters=jest.fn( async ()=> { 
                ride.emit('pair-start',)
                ride.emit('pair-error','2')
                return false
            })
            ride.getAdapters= jest.fn().mockReturnValue( 
                [ { udid:'1', 
                    capabilties:[IncyclistCapability.Control, IncyclistCapability.Power, IncyclistCapability.Speed, IncyclistCapability.Cadence], 
                    adapter: {  
                        isStarted:jest.fn().mockReturnValue(false),
                        pause:jest.fn().mockResolvedValue(true)
                    }
                  },
                  { udid:'2', 
                    capabilties:[IncyclistCapability.Power], 
                    adapter: {  
                        isStarted:jest.fn().mockReturnValue(false),
                        pause:jest.fn().mockResolvedValue(true)
                    }
                  }
                ])

            

            let updates: Array<PairingState> = [];
            let numLogs = 0;
            // we are expecting 3 status updates, the last one to contain the 
            const res = await new Promise (done => {
                svc.on('log',()=>{ 
                    ++numLogs
                    if (numLogs===2) 
                        done(updates)
                })                   
                svc.start( (status:PairingState)=>{ 
                    updates.push(status)
                })
            })

            //expect(res).toMatchSnapshot()        
            //console.log(JSON.stringify(res,undefined,2))

            expect(logEvent).toHaveBeenNthCalledWith( 1,expect.objectContaining( {message:'Start Pairing'}))
            expect(access.scan).not.toHaveBeenCalled()
            expect(ride.startAdapters).toHaveBeenCalled()

            expect(svc.getState().canStartRide).toBe(true)
            expect(svc.getState().capabilities?.find(c=>c.capability===IncyclistCapability.Control)).toMatchObject({deviceName:'Ant+FE 1234', connectState:'connecting'})
            expect(svc.getState().capabilities?.find(c=>c.capability===IncyclistCapability.Power)).toMatchObject({deviceName:'Ant+FE 1234', connectState:'connecting'})
            expect(svc.getState().capabilities?.find(c=>c.capability===IncyclistCapability.Speed)).toMatchObject({deviceName:'Ant+FE 1234', connectState:'connecting'})
            expect(svc.getState().capabilities?.find(c=>c.capability===IncyclistCapability.Cadence)).toMatchObject({deviceName:'Ant+FE 1234', connectState:'connecting'})
            expect(svc.getState().capabilities?.find(c=>c.capability===IncyclistCapability.HeartRate)).toMatchObject({deviceName:undefined})
            expect(svc.getState).toMatchSnapshot()

            })

            afterEach( ()=>{
                svc.stop()
            })
        })
        describe('stop',()=>{})
        describe('startDeviceSelection',()=>{})
        describe('stopDeviceSelection',()=>{})
        describe('selectDevice',()=>{})
        describe('deleteDevice',()=>{})
        describe('changeInterfaceSettings',()=>{})
    })

    describe('helpers',()=>{
        describe('mapCapability',()=>{
            test('success',()=>{
                const devices = [
                    {udid: "f594f6a6-6ff6-4717-b980-b491002fd0d4", name: "Jetblack FE 2606", interface: "ant",selected:true },
                    {udid: "5a81bd7a-4fe3-40b7-8206-8b28cb5a44e7", interface: "ble", name: 'Jetblack Volt',selected:false}
                ];
                const c:CapabilityInformation ={                
                        capability: IncyclistCapability.Control,
                        devices,
                        disabled: false                  
                }
    
                const res = mappedCapability(c)
    
                expect(res).toMatchObject({
                    capability:IncyclistCapability.Control,
                    deviceName:'Jetblack FE 2606',
                    deviceNames: 'Jetblack FE 2606;Jetblack Volt',
                    disabled: false,
                    selected: "f594f6a6-6ff6-4717-b980-b491002fd0d4",
                    interface: 'ant',
                    devices
                })
            })
        })
    
    })


    describe('protected',()=>{
        describe('disableInterfaceInCapabilities',()=>{
            const ctrl = [
                {udid: "1", interface: "ant", name: "A", selected:true },
                {udid: "2", interface: "ble", name: 'B',selected:false},
            ];
    
            const pwr = [
                {udid: "3", interface: "ble", name: 'C',selected:true}
            ];
            const hrm = [
                {udid: "4", interface: "ant", name: 'D',selected:true}
            ];
    
    
            const capabilities = [
                { capability:'control',deviceName:'A', deviceNames: 'A,B', disabled: false, selected: "1", interface: 'ant', devices:ctrl, value:1},
                { capability:'power',deviceName:'C', deviceNames: 'C', disabled: false, selected: "3", interface: 'ble', devices:pwr,value:2,},
                { capability:'heartrate',deviceName:'D', deviceNames: 'D', disabled: false, selected: "4", interface: 'ant', devices:hrm, value:3}
            
            ]
            let svc:TestWrapper
            beforeEach( ()=>{
                svc = new TestWrapper()
                svc._set('isInterfaceEnabled', jest.fn().mockReturnValue(true))            
                svc._set('state', {capabilities:clone(capabilities)})
                svc._set('access', { disableInterface:jest.fn()}  )
                svc._set('configuration',{ select:jest.fn(), unselect:jest.fn()})
            })
    
            test('active device has interface',()=>{
                const changed = svc._call('disableInterfaceInCapabilities','ant')           
    
                //const changed = (svc as any).disableInterfaceInCapabilities('ant')           
                const res = svc._get('state').capabilities
                
    
                expect(changed).toBe(true)
                expect( res[0].selected).toBe('2')
                expect( res[1].selected).toBe('3')
                expect( res[2]).toMatchObject( {selected:undefined, deviceName:undefined, value:undefined, interface:undefined})
    
                expect( res[0].devices[0].selected).toBe(false)
                expect( res[0].devices[1].selected).toBe(true)
                expect( res[1].devices[0].selected).toBe(true)
                expect( res[2].devices[0].selected).toBe(false)
    
            })
    
            test('no active device has interface',()=>{
                const changed = svc._call('disableInterfaceInCapabilities','ble')
                const res = (svc as any).state.capabilities
                
                expect(changed).toBe(true)
                expect( res[0].selected).toBe('1')
                expect( res[1]).toMatchObject( {selected:undefined, deviceName:undefined, value:undefined, interface:undefined})
                expect( res[2].selected).toBe('4')
    
                expect( res[0].devices[0].selected).toBe(true)
                expect( res[0].devices[1].selected).toBe(false)
                expect( res[1].devices[0].selected).toBe(false)
                expect( res[2].devices[0].selected).toBe(true)
            })
    
            test('interface is disabled',()=>{
    
                svc._set('isInterfaceEnabled', jest.fn().mockReturnValue(false))
                const changed = svc._call('disableInterfaceInCapabilities','ant')
                const res = (svc as any).state.capabilities
                
                expect(changed).toBe(true)
                expect( res[0]).toMatchObject( {selected:undefined, deviceName:undefined, value:undefined, interface:undefined})
                expect( res[1].selected).toBe('3')
                expect( res[2]).toMatchObject( {selected:undefined, deviceName:undefined, value:undefined, interface:undefined})
    
                expect( res[0].devices[0].selected).toBe(false)
                expect( res[0].devices[1].selected).toBe(false)
                expect( res[1].devices[0].selected).toBe(true)
                expect( res[2].devices[0].selected).toBe(false)
            })
    
            test('no device impacted',()=>{
                const changed = svc._call('disableInterfaceInCapabilities','serial')
                expect(changed).toBe(false)
            })
    
    
            
        })
    
     
        describe('enableInterfaceInCapabilities',()=>{
            const ctrl = [
                {udid: "1", interface: "ant", name: "A", selected:true },
                {udid: "2", interface: "ble", name: 'B',selected:false},
            ];
    
            const pwr = [
                {udid: "3", interface: "ble", name: 'C',selected:false}
            ];
            const hrm = [
                {udid: "4", interface: "ant", name: 'D',selected:true}
            ];
            const cad = [
                {udid: "5", interface: "ant", name: 'E',selected:true, connectState:'failed'},
                {udid: "6", interface: "ble", name: 'F',selected:false}
            ];
            const spd = [
                {udid: "7", interface: "ant", name: 'G',selected:false},
                {udid: "8", interface: "ble", name: 'H',selected:false}
            ];
    
    
            const capabilities = [
                { capability:'control',deviceName:'A', deviceNames: 'A;B', disabled: false, selected: "1", interface: 'ant', devices:ctrl, value:1},
                { capability:'power',disabled: false, devices:pwr,value:2,},
                { capability:'heartrate',deviceName:'D', deviceNames: 'D', disabled: false, selected: "4", interface: 'ant', devices:hrm, value:3},
                { capability:'cadence',deviceName:'E', deviceNames: 'E;F', disabled: false, selected: "5", interface: 'ant', devices:cad, value:90},
                { capability:'speed',disabled: false, devices:spd}
            
            ]
            let svc:TestWrapper;
            beforeEach( ()=>{
                svc = new TestWrapper();
                svc._set('isInterfaceEnabled', jest.fn().mockReturnValue(true))
                svc._set('state',{capabilities:clone(capabilities)})
                svc._set('access', { disableInterface:jest.fn()});
                svc._set('configuration', { select:jest.fn(), unselect:jest.fn()});         
            })
    
            test('single device not selected',()=>{
                const changed = svc._call('enableInterfaceInCapabilities','ble')           
                const res = (svc as any).state.capabilities
    
                expect(changed).toBe(true)
                expect( res[1].selected).toBe('3')
                expect( res[1].devices[0].selected).toBe(true)
    
            })
    
            test('multiple devices none selected',()=>{
                const changed = svc._call('enableInterfaceInCapabilities','ble')           
                const res = (svc as any).state.capabilities
    
                expect(changed).toBe(true)
                expect( res[4].selected).toBe('8')
                expect( res[4].devices[1].selected).toBe(true)
            })
    
            test('multiple devices, prev selected in failed state',()=>{
                const changed = svc._call('enableInterfaceInCapabilities','ble')           
                const res = (svc as any).state.capabilities
    
                expect(changed).toBe(true)
                expect( res[3]).toMatchObject( {selected:'6', deviceName:'F'})
                expect( res[3].devices[0].selected).toBe(false)
                expect( res[3].devices[1].selected).toBe(true)
            })
    
            test('no device impacted',()=>{
                const changed = svc._call('enableInterfaceInCapabilities','serial')           
                expect(changed).toBe(false)
            })
    
    
            
        })
    })



} )