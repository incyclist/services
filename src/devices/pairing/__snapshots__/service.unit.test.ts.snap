// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PairingService api start FE device in configuration 1`] = `[Function]`;

exports[`PairingService api start no devices in configuration will start scan 1`] = `
[
  {
    "canStartRide": false,
    "capabilities": [
      {
        "capability": "control",
        "deviceNames": "",
        "devices": [],
        "disabled": false,
      },
      {
        "capability": "power",
        "deviceNames": "",
        "devices": [],
        "disabled": false,
      },
      {
        "capability": "cadence",
        "deviceNames": "",
        "devices": [],
        "disabled": false,
      },
      {
        "capability": "speed",
        "deviceNames": "",
        "devices": [],
        "disabled": false,
      },
      {
        "capability": "heartrate",
        "deviceNames": "",
        "devices": [],
        "disabled": false,
      },
      {
        "capability": "bike",
        "deviceNames": "",
        "devices": [],
        "disabled": false,
      },
    ],
    "interfaces": [
      {
        "enabled": true,
        "isScanning": false,
        "name": "ant",
        "state": "connected",
      },
      {
        "enabled": true,
        "isScanning": false,
        "name": "ble",
        "state": "connected",
      },
      {
        "enabled": false,
        "isScanning": false,
        "name": "serial",
        "protocol": "Daum Classic",
        "state": "connected",
      },
      {
        "enabled": true,
        "isScanning": false,
        "name": "tcpip",
        "state": "connected",
      },
    ],
  },
  {
    "capabilities": [
      {
        "capability": "control",
        "deviceName": undefined,
        "deviceNames": "",
        "devices": [],
        "disabled": false,
        "interface": undefined,
        "selected": undefined,
      },
      {
        "capability": "power",
        "deviceName": undefined,
        "deviceNames": "",
        "devices": [],
        "disabled": false,
        "interface": undefined,
        "selected": undefined,
      },
      {
        "capability": "cadence",
        "deviceName": undefined,
        "deviceNames": "",
        "devices": [],
        "disabled": false,
        "interface": undefined,
        "selected": undefined,
      },
      {
        "capability": "speed",
        "deviceName": undefined,
        "deviceNames": "",
        "devices": [],
        "disabled": false,
        "interface": undefined,
        "selected": undefined,
      },
      {
        "capability": "heartrate",
        "deviceName": undefined,
        "deviceNames": "",
        "devices": [],
        "disabled": false,
        "interface": undefined,
        "selected": undefined,
      },
      {
        "capability": "bike",
        "deviceName": undefined,
        "deviceNames": "",
        "devices": [],
        "disabled": false,
        "interface": undefined,
        "selected": undefined,
      },
    ],
  },
]
`;

exports[`PairingService api start no devices in configuration; FE device found in scan 1`] = `[Function]`;

exports[`PairingService api start no devices in configuration; HRM sensor found in scan 1`] = `[Function]`;
